src/Xrefcheck/Verify.hs

```diff
@@ -81,230 +87,324 @@ toVerifyRes = VerifyResult . either one (\() -> [])
+            Left (url, option) ->
+              runReq defaultHttpConfig $
+              req method url NoReqBody ignoreResponse option
+            Right (url, option) ->
+              runReq defaultHttpConfig $
+              req method url NoReqBody ignoreResponse option
+
+      let maxTime = Time @Second $
+            unTime vcExternalRefCheckTimeout * timeoutFrac
+
+      mres <- liftIO (timeout maxTime $ void reqLink) `catch`
+        (either throwError (\() -> return (Just ())) . interpretErrors)
+      maybe (throwError $ ExternalResourceSomeError "Response timeout") pure mres
 
     isAllowedErrorCode = or . sequence
-        -- We have to stay conservative - if some URL can be accessed under
-        -- some circumstances, we should do our best to report it as fine.
-        [ (403 ==)  -- unauthorized access
-        , (405 ==)  -- method mismatch
-        ]
+      -- We have to stay conservative - if some URL can be accessed under
+      -- some circumstances, we should do our best to report it as fine.
+      [ (403 ==)  -- unauthorized access
+      , (405 ==)  -- method mismatch
+      ]
 
     interpretErrors = \case
-        JsonHttpException _ -> error "External link JSON parse exception"
-        VanillaHttpException err -> case err of
-            InvalidUrlException{} -> error "External link URL invalid exception"
-            HttpExceptionRequest _ exc -> case exc of
-                StatusCodeException resp _
-                    | isAllowedErrorCode (statusCode $ responseStatus resp) -> Right ()
-                    | otherwise -> Left $ ExternalResourceUnavailable (responseStatus resp)
-                other -> Left . ExternalResourceSomeError $ show other
+      JsonHttpException _ -> error "External link JSON parse exception"
+      VanillaHttpException err -> case err of
+        InvalidUrlException{} -> error "External link URL invalid exception"
+        HttpExceptionRequest _ exc -> case exc of
+          StatusCodeException resp _
+            | isAllowedErrorCode (statusCode $ responseStatus resp) -> Right ()
+            | otherwise -> Left $
+                ExternalHttpResourceUnavailable (responseStatus resp)
+          other -> Left . ExternalResourceSomeError $ show other
+
+    checkFtp :: URI -> Bool -> ExceptT VerifyError IO ()
+    checkFtp uri secure = do
+      -- get authority which stores host and port
+      authority <- case uriAuthority uri of
+        Right a -> pure a
+        Left _ -> throwError ExternalResourceInvalidUri
+      let host = T.unpack . unRText $ authHost authority
+      let port :: Int = fromIntegral . fromMaybe 21 $ authPort authority
+      let pieces = uriPath uri
+      -- in case there aro no pieces - smth wrong with url
+      unless (isJust pieces) $
+        throwError ExternalResourceInvalidUri
+      let (trailing, pathList) = fromJust pieces
+      -- in case there are trailing slash - smth wrong with url
+      when trailing $
```

Comment:
Is it really necessary to exclude a trailing slash?

Reply:
if there are trailing slash - this is directory, not a file

Reply:
I think it's fine to have a link to an FTP directory though, right?

Reply:
Hw.. you can open this link in your browser
ftp://statgen.ncsu.edu/pub/thorne/molevoclass/AtchleyOct19.pdf

but you can't open this one ftp://statgen.ncsu.edu/pub/thorne/molevoclass/ - it's required to open external app to connect to this server

Reply:
I can open `ftp://statgen.ncsu.edu/pub/thorne/molevoclass/` in my file explorer, which shows a folder with files in it.

Reply:
okay, if you insist I'll delete this check)

Reply:
Thanks :)

We should also add a couple of tests for ftp/ftps links to 1) an empty directory and 2) a non-empty directory

Reply:
What we suppose to do in case of empty dir? Is it valid case for us?

Reply:
I'd say so. If it's a valid FTP link (which it is), then we should consider it valid.

Reply:
I've added a test, but not sure it's working... I need ftp with empty dir to test it)